internal/metrics/prometheus/prometheus.go

```diff
@@ -0,0 +1,179 @@
+ * LICENSE file in the root directory of this source tree.
+ */
+
+package prometheus
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"log/slog"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/nginx/agent/v3/internal/bus"
+	"github.com/nginx/agent/v3/internal/metrics"
+)
+
+const (
+	defaultTarget  = "http://127.0.0.1:30882/metrics"
+	DataSourceType = "PROMETHEUS"
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6@v6.7.0 -generate
+//counterfeiter:generate -o mock_message_bus.go . MessageBusContract
+//go:generate sh -c "grep -v github.com/nginx/agent/v3/internal/metrics/prometheus mock_message_bus.go | sed -e s\\/prometheus\\\\.\\/\\/g > mock_message_bus_fixed.go"
+//go:generate mv mock_message_bus_fixed.go mock_message_bus.go
+type MessageBusContract interface {
+	Register(int, []bus.Plugin) error
+	DeRegister(plugins []string) error
+	Process(...*bus.Message)
+	Run()
+	Context() context.Context
+	GetPlugins() []bus.Plugin
+	IsPluginAlreadyRegistered(string) bool
+}
+
+// Used for deserializion of parsed Prometheus data.
+type Scraper struct {
+	bus bus.MessagePipeInterface
+	// previousCounterMetricValues map[string]DataPoint
+	cancelFunc context.CancelFunc
+}
+
+func NewScraper(mp bus.MessagePipeInterface) *Scraper {
+	return &Scraper{
+		bus: mp,
+	}
+}
+
+func (s *Scraper) Start(ctx context.Context, updateInterval time.Duration) error {
+	slog.Info("prometheus data source booting up")
+	ticker := time.NewTicker(updateInterval)
+
+	cancellableCtx, cancelFunc := context.WithCancel(ctx)
+	s.cancelFunc = cancelFunc
+
+	// This should be discovered dynamically.
+	targets := []string{defaultTarget}
+	if value, ok := os.LookupEnv("PROMETHEUS_TARGETS"); ok {
+		targets = strings.Split(value, ",")
+	}
+
+	for _, target := range targets {
+		resp, err := http.DefaultClient.Get(strings.TrimSpace(target))
+		if err != nil {
+			return fmt.Errorf("could not find prometheus endpoint %s: %w", target, err)
+		}
+
+		log.Printf("found prometheus endpoint %v", resp.Request.URL)
+	}
+
+	for {
+		select {
+		case <-cancellableCtx.Done():
+			return nil
+		case <-ticker.C:
+			for _, target := range targets {
+				entries := scrapeEndpoint(strings.TrimSpace(target))
+				msgs := []*bus.Message{}
+				for _, e := range entries {
+					msgs = append(msgs, e.ToBusMessage())
+				}
+				s.bus.Process(msgs...)
+			}
+		}
+	}
+}
+
+func (s *Scraper) Stop() {
+	slog.Info("prometheus data source shutting down")
+	s.cancelFunc()
+}
+
+func (s *Scraper) Type() string {
+	return DataSourceType
+}
+
+func scrapeEndpoint(target string) []metrics.DataEntry {
+	resp, err := http.Get(target)
+	if err != nil {
+		slog.Error("GET to Prometheus HTTP scrape endpoint failed", "endpoint", target)
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		slog.Warn("failed to parse Prometheus scrape endpoint response body")
```

Comment:
Should include `err` value in log and maybe return early? Nothing to try to parse if it cannot be scraped

```suggestion
		slog.Warn("failed to parse Prometheus scrape endpoint response body", "error", err)
		return
```

Reply:
Fixed this in the refactor, good catch!